docs/proposal/2022-11-22-botkube-plugins-integration.md

```diff
@@ -0,0 +1,498 @@
+```yaml
+executors:
+  'plugin-based':
+    botkube/kubectl:     # <repo>/<plugin> is syntax for plugin based executors
+      enabled: true      # if not enabled we don't download and start a given plugin
+      version: v1.0.0    # if empty, the latest version is used.
+      config:            # plugin specific configuration
+      # ...
+sources:
+  'plugin-based':
+    botkube/kubernetes:
+      enabled: true
+      version: v1.0.0
+      config:
+      # ...
+```
+
+In this way you can:
+
+1. Register multiple plugins repositories.
+2. Enable different plugin version from the same repository.
+3. Enable the same plugin (e.g. `kubectl`) from different repositories.
+
+### Defining executors aliases
+
+Currently, the aliases are built-in into Botkube core logic. One option is to define them in the index file for a given executor:
+
+```yaml
+entries:
+  - name: "kubectl"
+    type: "executor"
+    description: "Kubectl executor plugin."
+    version: "v1.0.0"
+    aliases: [ "kc", "k" ]
+    links:
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/executor_kubectl-darwin-amd64
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/executor_kubectl-darwin-arm64
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/executor_kubectl-linux-amd64
+      - https://github.com/kubeshop/botkube/releases/download/v0.17.0/executor_kubectl-linux-arm64
+```
+
+Optionally we can allow to override them when enabling a given plugin:
+
+```yaml
+executors:
+  'plugin-based':
+    botkube/kubectl:
+      enabled: true
+      version: v1.0.0
+      aliases: [ "kc", "k" ]
+```
+
+### Validating plugin configuration
+
+We introduce a basic validation to make sure that:
+
+1. Entries define in the index files don't conflict with each other - having the same `name`, `type` and `version`:
+
+   <details>
+     <summary>Example</summary>
+
+   ```yaml
+   entries:
+     - name: "kubectl"
+       type: "executor"
+       description: "Kubectl executor plugin."
+       version: "v1.0.0"
+       links:
+         - https://github.com/kubeshop/botkube/releases/download/v0.18.0/executor_kubectl-darwin-amd64
+     - name: "kubectl"
+       type: "executor"
+       description: "Kubectl executor plugin."
+       version: "v1.0.0"
+       links:
+         - https://github.com/kubeshop/botkube/releases/download/v0.17.0/source_kubernetes-darwin-amd64
+   ```
+   </details>
+
+2. Enabled executors don't conflict with each other:
+   <details>
+     <summary>Example</summary>
+
+   ```yaml
+   executors:
+     'plugin-based':
+       botkube/kubectl:
+         enabled: true
+         config:
+           namespaces:
+             include: ["botkube", "default","ambassador"]
+           commands:
+             verbs: ["get","logs","top"]
+             resources: ["pods","deployments","nodes","configmap"]
+       mszostok/kubectl:  # not allowed as it is conflicting with already registered 'kubectl' from the Botkube repository.
```

Comment:
In this config I understand there is a conflict, but if I change the aliases I can have these two executors, right? If not, why?

If we listed all the options inside the repository index (`kubectl`, `kc`, `k`) then I would be able to override them for `mszostok/kubectl` - e.g. define `aliases: [ "mkubectl", "mk", "mkc" ]`. WDYT?

In that way options no 3 and 4 would be no longer valid. I think that would be a nice feature, and also very convenient for development/testing purposes.

Reply:
Personally, I have two concerns:
- is it a real use case? How you would use it for testing?
- if we will specify also the "original" name then IMO it doesn't fit into `aliases` definition 🤔

Reply:
> is there a real use case for it? How you would use it for testing?

1. For example, I'd be able to fetch your forked kubectl executor to test your changes from PR (if they don't affect the whole Botkube) without touching the core which I have installed on Oracle Cloud.
2. Also, what if there are two different plugins, but with the same name? Like `https://github.com/travis-ci/gimme` and mszostok/gimme` (I know, stupid example, but that was the first binary name that came to my mind 😄) which do different things, but share the same name? There won't be a way to install and use them both, even if they come from different repos (!). I think this is a serious issue.
3. I would also like to invert the question: why two separate binaries could run into a conflict which we cannot resolve? If I have two versions of kubectl, I can rename the binary / make an alias and run it. Same with Go / Node.js versions. I just don't see any reason why we would need to block downloading two plugins with the same name. If there's any obstacle to do it, then of course we should reevaluate this 👍 But enforcing the community having unique names across all repositories is a huge downside.

> if we will specify also the "original" name then IMO it doesn't fit into aliases definition 🤔

yep, we can come up with different name for the `aliases` property if you don't like it - `commands`, `prefixes`, ...?
But I think `aliases` is rather OK. IMHO name is just... a name, which can be different from the actual command. For example, I have `mongo-seeding` project, but CLI is called `seed`, because `mongo-seeding` is not really a great name for a binary.

Reply:
Let's discuss it f2f:

1.  This still won't be possible 🤔 because we don't publish such binaries. And event if that will be the case, then probably even easier will be to just change `botkube/kubectl` to `mszostok/kubectl` 🤔 
2. Yeah, but is it real thing, when sth like that will happen 😄 ?
3. I'm trying to introduce as little as possible when it comes to the new features. 
4. Even if we will have what you describe, then validation is still needed as user still can define the same `aliases` in two places.

TL;DR;
1. By default, we use the plugin name, e.g. `botkube/kubectl` resolves to `kubectl`
2. If user defines in the index file the`aliases/commands` property, then we use only them.
    ```yaml
      - name: "kubectl"
        type: "executor"
        description: "Kubectl executor plugin."
        version: "v1.0.0"
        aliases: [ "kc", "k" ]  # overrides under which names the plugin will be available
    ```
3. If users want, they can also override it once enabling:
    ```yaml
    executors:
      'plugin-based':
        botkube/kubectl@v1.0.0:
          enabled: true
          aliases: ["kubectl", "kc", "k"]   # overrides under which names the plugin will be available
          config: {}
    ```

We would need to have some mechanism to tell the end-user under which names the plugin will be available.

Consider such case:
1. I enable `botkube/kubectl@v1.0.0`
2. In the `botkube` repo index the `kubectl@v1.0.0` has the `aliases: [ "kc", "k" ]` property.

In this way, the enabled plugin will be available only under `kc` and `k`. We would need to somehow tell about that our users.

Reply:
What if admin is always required to define aliases? And the repo index won't contain them at all? Users will be in full control what prefixes they want to use. Why would a plugin author define aliases actually? 🤔

And yes, I didn't say we don't need validation for it 👍 it's more about name conflicts. I think if we can easily avoid them, let's do it 👍

Reply:
>What if admin is always required to define aliases? And the repo index won't contain them at all? Users will be in full control what prefixes they want to use. Why would a plugin author define aliases actually? 🤔

The only reason is that there are some well known aliases, e.g. for `kubectl`. Defining them in the index file solves one problem of redefining them each time you define configuration.

You can already see that we have multiple configuration for `kubectl` executor. If we do not have aliases defined in index file. Then each time we will need to remember to do that.